### DISCUSSION TITLE
Online stitching camera frames

### DISCUSSION THREAD

**Discussion by horsto** [bonsai-rx/bonsai#1090] [Q&A] [answered]
I have three USB cameras capturing video data of the same frame size and at the same speed. They are just tiled images that have an x,y offset, but no angular offset. I want to stitch them into one big (flat) composite (mosaic) and continue to work only with that. Downstream of that I want to run object detection on the stitched frame. 

The idea is similar to [this entry](https://groups.google.com/g/bonsai-users/c/mWYATk5vgz4/m/xqZkqzAhBAAJ). 

I am running into out of memory errors quite quickly after starting the workflow. But it does not seem like a lot of memory has been used / is requested. Any ideas? 
[test_stitching.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/9805271/test_stitching.bonsai.zip)

<img width="459" alt="Screenshot 2022-10-17 200227" src="https://user-images.githubusercontent.com/7548928/196305738-d2097570-7783-4dbe-b335-4064b6a5311a.png">

**Comment by bruno-f-cruz** [comment]
Hey, 

I never came across this error before in this pipeline. I wonder if your computer just can't keep up with the online stitching.
Can you describe the setup? What is the FPS of each camera, are they synchronized, Color vs Mono, and resolution?

If you can afford it, I would probably convert everything to grayscale and use `WithLatestFrom` instead of `CombineLatest`.

**Reply by glopesdev** [reply]
It's likely the CombineLatest that is the problem indeed. Assuming a simple 1 MP image at 60Hz from one camera, then CombineLatest > Concat would be allocating 3 MP images at 180Hz.

**Reply by horsto** [reply]
Thanks! Indeed the images are very large (see screenshot), acquiring at 20fps each. This might be too much data? 
The `Concat` node seems to throw the error- no errors when concat is skipped it seems. 
Even streams from two cameras are too much (see screenshot). I wonder why though. Conversion to Gray is not an option since I need the full color information. 

<img width="538" alt="with_latest_from1" src="https://user-images.githubusercontent.com/7548928/196447563-8b555c6d-1d89-47ac-b84b-ed3a0f9841b8.png">
<img width="1468" alt="combine_latest_1" src="https://user-images.githubusercontent.com/7548928/196447565-a97a9373-8ae5-4aec-9557-f5b62212faca.png">

**Comment by horsto** [comment]
Grayscale conversion + `WithLatestFrom` + grayscale gives me some weirdness on the camera side (so maybe the memory error isn't telling the whole truth ... ) 


<img width="1313" alt="Screenshot 2022-10-18 101245" src="https://user-images.githubusercontent.com/7548928/196455247-21f1ddbe-1ed1-46d6-ae64-fc1183c08e9e.png">

**Comment by horsto** [comment, marked as answer]
This actually solved itself. Turns out I was starting `Bonsai(x86)` instead of `Bonsai` - The latter has zero issues with frame grabbing + stitching.
